### ISSUE TITLE
Issue about mysql,  packet handler memory,  room list

### ISSUE THREAD

**Issue by jl9404** [Cleroth/sp_server#16]
First of all,  the mysql_use_result function will return null or even return 'cannot access memory' and it will lead the server crash.  I have tried to increase the mysql server buffer and memory,  and change the table engine to Myisam,  it still return null or error,  it will be horrible.

For the memory issue,  sometimes the packethandler memory cannot be read,  so it will crash the server.  I am thinking about will it related to the handling disconnect from user?

The room capacity is fixed to 22, and the server may crash if u change to any other number,  it seems the packet format is not that precise?

**Comment by tsuncp** [comment]
For memory issues
I suggest run the channel server from visual studio
So that when it happen again , you can log everything

jl9404 <notifications@github.com>於 2016年2月24日 週三，上午10:03寫道：

**Comment by Cleroth** [comment]
That's part the problem with using MySQL C bindings, it's too easy to shoot yourself in the foot, and the error handling is really poor.
Have you tried looking at how much memory the channel server and the mysqld is using when it crashes? I imagine the PacketHandler may be leaking memory, which eventually causes it to crash.

**Comment by Cleroth** [comment]
Also, you don't need to run it from VS, you can attach VS when it crashes (provided you have the .pdb in the folder).

**Comment by tsuncp** [comment]
because when I run out of vs . the channel server will shutdown without any warning

**Comment by Cleroth** [comment]
You probably have the error reporting disabled.
> On Windows 7, go to “Control Panel\System and Security\Action Center\Change Action Center settings\Problem Reporting Settings” and select "Each time a problem occurs, ask me before checking for solutions".

**Comment by jl9404** [comment]
how to view the .pdb file
i am new to visual studio lol
btw, packethandler always has memory leakage...

**Comment by tsuncp** [comment]
When you build the binary file. You will have three files. 

Channl32_2015_d.exe Channl32_2015_d.ilk Channl32_2015_d.pdb

put them together, when the program is crash , you can use VS to debug.

**Comment by Cleroth** [comment]
Yea, I thought memory leaks would be the cause. But they won't crash the program (generally) unless it runs out of memory.

**Comment by tsuncp** [comment]
i dun think access memory error is caused by memory leak. If runs out of memory, the error must show " out of memory " message

**Comment by Cleroth** [comment]
It's possible mysql ran out of memory and just responds with with empty set.
There seems to be several crashing bugs though, so that's why I'm asking. I'm just ruling out possible causes.

**Comment by tsuncp** [comment]
I am not challenging you loooool. i just want to exclude this reason.

**Comment by jl9404** [comment]
it seems that i use memset to clear the Rooms[i].Player[i] instead of just set it to "0" would fix the memory issue

**Comment by Cleroth** [comment]
What do you mean? They're all being initialized to 0 `_RoomList`'s constructor.

**Comment by jl9404** [comment]
i know i changed the function Exitplayer in userlist.h
```
memset(&Rooms[i].Player[j], 0, sizeof(Rooms[i].Player[j]));
memset(&Rooms[i].users[j], 0, sizeof(Rooms[i].users[j]));
```
the packethandler memory leakage issue will be away

**Comment by whitearies** [comment]
This error happens when player exit room or game abnormaly, and mysql will send back empty data.
Add these code before “MYSQL_RES *res = mysql_use_result(connection);” should help.
```
    iif (mysql_query(connection, buffer))
    {
        cout << "NO DATA" << endl;
        return;
    }
```

**Comment by Cleroth** [comment]
There's no need for memset. Just set it to 0 on declaration.

**Comment by whitearies** [comment]
I use this to avoid mysql_query  error because of data = NULL.
    mysql_query(connection, buffer);
    MYSQL_RES *res = mysql_use_result(connection);
    if (res == NULL || !res->row_count)
    {
        cout << "NO DATA" << endl;
        return;
    }
    MYSQL_ROW result = mysql_fetch_row(res);

**Comment by Cleroth** [comment]
if (!res)
	{
		cout << "MySQL error: " << mysql_error(connection) << endl;
		return;
	}

**Comment by jl9404** [comment]
i dont why lol, after i use memset the error is gone...

**Comment by Cleroth** [comment]
You increased the number of rooms or players?

**Comment by jl9404** [comment]
no i cant i have tried before 
however it is not working
i changed the packet size, the array size, and the lobby max room

**Comment by whitearies** [comment]
`if (!res)` Dosen't work when you get NULL.

**Comment by jl9404** [comment]
i may update the mysql c lib to the updated one as the memory error throw from the mysql lib

**Comment by Cleroth** [comment]
NULL is defined as 0. `if(!res)` does exactly the same as `if(res == NULL)`

**Comment by Cleroth** [comment]
And NULL is Windows garbage that should never be used.

**Comment by Cleroth** [comment]
mysql lib throwing a memory error doesn't necessarily mean it's the library's fault, but the usage could be wrong.

**Comment by jl9404** [comment]
do u know how to increase the room limit?
although i edit the packet size, array size, and lobby max room
it does not work

**Comment by Cleroth** [comment]
Where does it crash if you increase it?

**Comment by Cleroth** [comment]
I've updated the MySQL library, try it out.

**Comment by jl9404** [comment]
`PackHandle.GenerateResponse(ROOM_LIST_RESPONSE);`

`struct RoomListResponse
{
	int size; //0xC90
	int type;
	int unk1; //11036
	int checksum;
	int state;
	int roomnumber[200];
	char title[200][29];
	short zero1;
	int mode[200];
	int map[200];
	int unks2[200]; // current player
	int maxplayers[200];
	int zeros[5];
	short szero;
	bool bPotion[200];
	bool bClosed[200];
	short szero1;
	int unk3[11];
	int unks4[200]; //-1,-1
	int players[200][8]; //0x8C
	int unks6[291];
	short zero2;
	bool bunk[200]; //1..
	int zeros1[5];
	int unk4; // 0x500000
};`

i changed the packet struct like this am i correct?

**Comment by Cleroth** [comment]
You can't change packet structure. The client won't recognize the new packet structure.

**Comment by jl9404** [comment]
so there is no way to change the room limit ?

**Comment by Cleroth** [comment]
Unlikely.

**Comment by jl9404** [comment]
As i just saw a server from china, they increase the room limit to 200 lol.

**Comment by Cleroth** [comment]
Either they modified the client, or there's some way the server tells the client how many maximum rooms there are.

**Comment by jl9404** [comment]
yes there is a packet to specific the maximum room
`Join_Channel_PlayerData_Response.unk44 = 22; //0x108`

**Comment by whitearies** [comment]
It is able to change maxroom , we have found it before

**Comment by whitearies** [comment]
Join_Channel_PlayerData_Response.unk44 = 22; //Maxroom
and replace all "< 22" as "< max_room_number"
and  LobbyRoom Rooms[max_room_number];

**Comment by Cleroth** [comment]
Yea, be sure to add a way to change it easily. Magic numbers are devil.

**Comment by jl9404** [comment]
how about the packet struct? no need to change?

**Comment by Cleroth** [comment]
It should be changed.

**Comment by jl9404** [comment]
@whitearies 
can u provide more details please
thank you

**Comment by whitearies** [comment]
Just change what I said, that would work.

**Comment by whitearies** [comment]
I suggest to put the max_room_number into config files.

**Comment by whitearies** [comment]
There is an error when I merged:
Fatal error: LNK1181: cannot open input file 'libmysql.lib'

**Comment by Cleroth** [comment]
Yea, the libs were excluded. I just pushed them.

**Comment by Cleroth** [comment]
From config files wouldn't be constant, so the code would need to change (arrays need to be constant).

**Comment by jl9404** [comment]
oh , it does not work...

**Comment by Cleroth** [comment]
Join the gitter: https://gitter.im/Cleroth/sp_server

**Comment by Cleroth** [comment]
result bug fixed by 11fa76b97704a7bb2479e0efa5a1db8d5164adb2
For empty result sets, each query that may have no results need checks.
